### ISSUE TITLE
NVP progress

### ISSUE THREAD

**Issue by ntustison** [ntustison/ANTsXNetTraining#3]
Hey @stnava,

Posting an issue here so you know where I'm [putting the NVP work temporarily](https://github.com/ntustison/ANTsXNetTraining/tree/main/NVP).  So this code appears to be working---see below.  I had to make some changes but I'm pretty pleased with the progress so far.  I had to make some changes in terms of modifying it so that it works with images, images with masks, and vectors.  

I'm going to let it run for a bit on the set of 2-D axial T1 slices I put together for the in-painting network (n=~40000).  Then I'll test to see if we can actually do the inverse transform to interpolate the learned distribution model to create new images. Once I get that working, I'll get back to you and we can decide next steps.

```
2025-02-18 11:42:19.389872: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2025-02-18 11:42:19.422436: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI AVX512_BF16 AVX_VNNI AMX_TILE AMX_INT8 AMX_BF16 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2025-02-18 11:42:19.894289: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
Total number of training brain slices:  39280
2025-02-18 11:42:21.121019: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1639] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 46125 MB memory:  -> device: 0, name: NVIDIA RTX A6000, pci bus id: 0000:ac:00.0, compute capability: 8.6
/home/ntustison/anaconda3/envs/antsx/lib/python3.11/site-packages/tensorflow_probability/python/bijectors/real_nvp.py:392: UserWarning: `tf.layers.dense` is deprecated and will be removed in a future version. Please use `tf.keras.layers.Dense` instead.
  x = tf1.layers.dense(
/home/ntustison/anaconda3/envs/antsx/lib/python3.11/site-packages/keras/src/legacy_tf_layers/base.py:623: UserWarning: `layer.updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  self.updates, tf.compat.v1.GraphKeys.UPDATE_OPS
/home/ntustison/anaconda3/envs/antsx/lib/python3.11/site-packages/tensorflow_probability/python/bijectors/real_nvp.py:398: UserWarning: `tf.layers.dense` is deprecated and will be removed in a future version. Please use `tf.keras.layers.Dense` instead.
  x = tf1.layers.dense(
Loading /home/ntustison/Data/NVP/Scripts/nvp_t1_axial.h5
Epoch 1/1000
2025-02-18 11:44:38.328258: I tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:606] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
32/32 [==============================] - ETA: 0s - negative_log_likelihood: 84978.1753 - bits_per_dim: 1.8707   
Epoch 1: negative_log_likelihood improved from inf to 76691.03125, saving model to /home/ntustison/Data/NVP/Scripts/nvp_t1_axial.h5
32/32 [==============================] - 113s 1s/step - negative_log_likelihood: 84727.0497 - bits_per_dim: 1.8652 - lr: 2.0000e-04
Epoch 2/1000
32/32 [==============================] - ETA: 0s - negative_log_likelihood: 68344.0665 - bits_per_dim: 1.5045 
Epoch 2: negative_log_likelihood improved from 76691.03125 to 60567.29688, saving model to /home/ntustison/Data/NVP/Scripts/nvp_t1_axial.h5
32/32 [==============================] - 39s 1s/step - negative_log_likelihood: 68108.4068 - bits_per_dim: 1.4993 - lr: 2.0000e-04
Epoch 3/1000
32/32 [==============================] - ETA: 0s - negative_log_likelihood: 52902.5864 - bits_per_dim: 1.1646 
Epoch 3: negative_log_likelihood improved from 60567.29688 to 45382.67188, saving model to /home/ntustison/Data/NVP/Scripts/nvp_t1_axial.h5
32/32 [==============================] - 39s 1s/step - negative_log_likelihood: 52674.7102 - bits_per_dim: 1.1596 - lr: 2.0000e-04
Epoch 4/1000
32/32 [==============================] - ETA: 0s - negative_log_likelihood: 37833.0845 - bits_per_dim: 0.8328 
Epoch 4: negative_log_likelihood improved from 45382.67188 to 30639.61328, saving model to /home/ntustison/Data/NVP/Scripts/nvp_t1_axial.h5
32/32 [==============================] - 39s 1s/step - negative_log_likelihood: 37615.1005 - bits_per_dim: 0.8281 - lr: 2.0000e-04
Epoch 5/1000
32/32 [==============================] - ETA: 0s - negative_log_likelihood: 23341.0920 - bits_per_dim: 0.5138 
Epoch 5: negative_log_likelihood improved from 30639.61328 to 16605.30664, saving model to /home/ntustison/Data/NVP/Scripts/nvp_t1_axial.h5
32/32 [==============================] - 38s 1s/step - negative_log_likelihood: 23136.9773 - bits_per_dim: 0.5093 - lr: 2.0000e-04
Epoch 6/1000
32/32 [==============================] - ETA: 0s - negative_log_likelihood: 10068.0340 - bits_per_dim: 0.2216 
Epoch 6: negative_log_likelihood improved from 16605.30664 to 3937.38672, saving model to /home/ntustison/Data/NVP/Scripts/nvp_t1_axial.h5
32/32 [==============================] - 39s 1s/step - negative_log_likelihood: 9882.2568 - bits_per_dim: 0.2175 - lr: 2.0000e-04
Epoch 7/1000
32/32 [==============================] - ETA: 0s - negative_log_likelihood: -2535.2205 - bits_per_dim: -0.0558 
Epoch 7: negative_log_likelihood improved from 3937.38672 to -8198.38477, saving model to /home/ntustison/Data/NVP/Scripts/nvp_t1_axial.h5
32/32 [==============================] - 39s 1s/step - negative_log_likelihood: -2706.8316 - bits_per_dim: -0.0596 - lr: 2.0000e-04
Epoch 8/1000

```
